### PR TITLE
[BEAM-495] Create General Verifier for File Checksum

### PR DESCRIPTION
**Pull request by markflyhigh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

JIRA(https://issues.apache.org/jira/browse/BEAM-495)

Generalize FileChecksumMatcher which is from WordCountIT. FileChecksumMatcher can be used in E2E test to verify checksum of output result.

### CODE REVIEW

**Comment by markflyhigh** [comment]
+R: @jasonkuster @dhalperi

**Comment by jasonkuster** [comment]
LGTM

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  @Mock
+  private PipelineResult pResult = Mockito.mock(PipelineResult.class);
+
+  @Test
+  public void testPreconditionValidChecksumString() throws IOException{
+    String tmpPath = tmpFolder.newFile().getPath();
+
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage(containsString("Expected valid checksum, but received"));
+    new FileChecksumMatcher(null, tmpPath);
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java]: this line will throw so that means that the line below will never get called

You need to either catch and verify yourself or split this into two tests, one with null and one with ""

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    new FileChecksumMatcher("", tmpPath);
+  }
+
+  @Test
+  public void testPreconditionValidFilePath() throws IOException {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage(containsString("Expected valid file path, but received"));
+    new FileChecksumMatcher("checksumString", null);
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java]: this line will throw so that means that the line below will never get called

You need to either catch and verify yourself or split this into two tests, one with null and one with ""

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      List<String> outputs = readLines(filePath);
+
+      // Verify outputs. Checksum is computed using SHA-1 algorithm
+      actualChecksum = hashing(outputs);
+      LOG.info("Generated checksum for output data: {}", actualChecksum);
+
+      return actualChecksum.equals(expectedChecksum);
+    } catch (IOException e) {
+      throw new RuntimeException(
+              String.format("Failed to read from path: %s", filePath));
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java]: 4 spaces not 8

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    IOChannelFactory factory = IOChannelUtils.getFactory(path);
+
+    // Match inputPath which may contains glob
+    Collection<String> files = factory.match(path);
+
+    // Read data from file paths
+    int i = 0;
+    for (String file : files) {
+      try (Reader reader =
+               Channels.newReader(factory.open(file), StandardCharsets.UTF_8.name())) {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java]: alignment issue, 4 spaces not 8 should preceed Channels.newReader

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    for (String str : strs) {
+      hashCodes.add(Hashing.sha1().hashString(str, StandardCharsets.UTF_8));
+    }
+    return Hashing.combineUnordered(hashCodes).toString();
+  }
+
+  @Override
+  public void describeTo(Description description) {
+    description
+            .appendText("Expected checksum is (")
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java]: 4 spaces not 8 here as well

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            .appendText(expectedChecksum)
+            .appendText(")");
+  }
+
+  @Override
+  public void describeMismatchSafely(PipelineResult pResult, Description description) {
+    description
+            .appendText("was (")
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java]: 4 spaces not 8 here as well

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    new FileChecksumMatcher("checksumString", "");
+  }
+
+  @Test
+  public void testChecksumVerify() throws IOException{
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java]: Would be useful if the test matched two or more files

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java (hunk 1 above)
Comment (markflyhigh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java (hunk 2 above)
Comment (markflyhigh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java (hunk 3 above)
Comment (markflyhigh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java (hunk 4 above)
Comment (markflyhigh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java (hunk 5 above)
Comment (markflyhigh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java (hunk 6 above)
Comment (markflyhigh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/FileChecksumMatcher.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java (hunk 7 above)
Comment (markflyhigh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java]: done by adding another test to matching two files.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java [hunk 8]
```diff
@@ -49,30 +49,58 @@
...
+  @Test
+  public void testMatcherVerifyMultipleFilesInOneDir() throws IOException {
+    File tmpFile1 = tmpFolder.newFile();
+    File tmpFile2 = tmpFolder.newFile();
+    Files.write("To be or not to be, ", tmpFile1, StandardCharsets.UTF_8);
+    Files.write("it is not a question.", tmpFile2, StandardCharsets.UTF_8);
+    FileChecksumMatcher matcher =
+        new FileChecksumMatcher(
+            "90552392c28396935fe4f123bd0b5c2d0f6260c8",
+            tmpFolder.getRoot().getPath() + "/*");
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java]: Use IOChannelUtils.resolve or equivalent, your assuming all filesystems use '/' as the separator

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java [hunk 9]
```diff
@@ -49,30 +49,58 @@
...
+
+    assertThat(pResult, matcher);
+  }
  }
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java]: ' }' -> '}'
remove the extraneous space

File: examples/java/src/test/java/org/apache/beam/examples/WordCountIT.java [hunk 10]
```diff
@@ -59,8 +40,13 @@
 
   /**
    * Options for the WordCount Integration Test.
+   *
+   * Define expected output file checksum to verify WordCount pipeline result with customized input.
```
Comment (lukecwik) [inline examples/java/src/test/java/org/apache/beam/examples/WordCountIT.java]: missing <p> tag for second paragraph

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java (hunk 8 above)
Comment (markflyhigh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java (hunk 9 above)
Comment (markflyhigh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/FileChecksumMatcherTest.java]: done. Good catch!!

File: examples/java/src/test/java/org/apache/beam/examples/WordCountIT.java (hunk 10 above)
Comment (markflyhigh) [inline examples/java/src/test/java/org/apache/beam/examples/WordCountIT.java]: done. Thanks!

**Comment by markflyhigh** [comment]
PTAL @lukecwik

**Comment by lukecwik** [comment]
LGTM
